### ISSUE TITLE
IndexedJob Restarts Succeeded Pods and Completion Takes Much Longer Than Expected

### ISSUE THREAD

**Issue by eddieparker** [kubernetes/kubernetes#108887]
### What happened?

I'm using the IndexedJob featuregate to run some ML training.  I spin up an indexed job with 10 parallelism & completions required. 

The code internally has a list of tests it wants to run which is < 10.  If it gets assigned an index from k8s that is higher then the number of tests, it early returns (sys.exit(0) in python), otherwise it does an ML training that takes about an hour.

What I notice happening is that the job itself starts marking completions (as expected) but then it randomly starts retrying already-succeeded indexes, wiping out them from the 'is completed' list on the job itself.  Because of this, and that my trainings stagger somewhat, the job never completes when I'd expect this job to take as long as the 'longest' individual task.

### What did you expect to happen?

I'd expect the longest individual task to be the entire length of the job (allowing they're all scheduled at the same time), and once complete they'd stay completed.

### How can we reproduce it (as minimally and precisely as possible)?

I made an example yaml that I believe shows this off.  It basically spins up an indexed job and makes each index wait N minutes, where N is the individual index.

```apiVersion: batch/v1
kind: Job
metadata:
  labels:
    app: test-indexed-job
  name: test-indexed-job
spec:
  completions: 5
  parallelism: 5
  completionMode: Indexed
  template:
    metadata:
      annotations:
        prometheus.io/scrape: "true"
      labels:
        app: test-indexed-job
    spec:
      containers:
        - image: busybox
          name: test-indexed-job
          securityContext:
            capabilities:
              add:
              - SYS_ADMIN
          resources:
            limits:
              cpu: "1"
              memory: 100Gi
              ephemeral-storage: 1Gi
            requests:
              cpu: "1"
              memory: 1Gi
              ephemeral-storage: 1Gi
          args:
            - /bin/sleep
            - $(JOB_COMPLETION_INDEX)m
      restartPolicy: Never
```

Running it locally and tuning back in after 10 minutes I see it's still running, even though each of the indices have completed multiple times by now:

![image](https://user-images.githubusercontent.com/1141487/159547150-43ada2b7-1afe-46ae-a577-48d7f8ea4104.png)

### Kubernetes version

<details>

```console
Client Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.5", GitCommit:"5c99e2ac2ff9a3c549d9ca665e7bc05a3e18f07e", GitTreeState:"clean", BuildDate:"2021-12-16T08:38:33Z", GoVersion:"go1.16.12", Compiler:"gc", Platform:"windows/amd64"}
Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.8", GitCommit:"4a3b558c52eb6995b3c5c1db5e54111bd0645a64", GitTreeState:"clean", BuildDate:"2021-12-15T14:46:22Z", GoVersion:"go1.16.12", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>
N/A: Bare metal.
</details>

### OS version

<details>

```console
BuildNumber  Caption                                    OSArchitecture  Version
19043        Microsoft Windows 10 Pro for Workstations  64-bit          10.0.19043
```

</details>

### Install tools

<details>
N/A
</details>

### Container runtime (CRI) and version (if applicable)

<details>
Not sure how to get this?  If guided I can try and provide.
</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>
Not sure how to get this?  If guided I can try and provide.
</details>

**Comment by eddieparker** [comment]
I'm not sure if this is the right sig/wg but please forward if I have it incorrect. 

/wg batch

**Comment by eddieparker** [comment]
/wg Batch

Maybe I had improper case?

**Comment by eddieparker** [comment]
Alright, take three. Stealing this from another issue, apologies if this is the wrong sig/area:

/sig apps
/area batch

**Comment by eddieparker** [comment]
Just in case it's useful, here's the output of the job when it finally completed.  Notable that it took ~80 minutes to finish, when I'd expect it to take ~5-6 minutes max.

![image](https://user-images.githubusercontent.com/1141487/159585678-95b5dd83-af8c-4495-927d-ce3ec10885bc.png)

**Comment by adriangb** [comment]
> This is expected in version 1.22

I hit this issue today. Just for my knowledge, when you say this is "expected" do you mean that it's a known issue or that this is the expected behavior? Is there any way to work around this on 1.22? In my case it looks like pods are getting deleted when the node gets removed (autoscaling).

**Comment by alculquicondor** [comment]
It is a know issue that we can't fix in 1.22.

The only workaround is to increase the terminated-pod-gc-threshold

**Comment by adriangb** [comment]
Gotcha, thank you for the quick reply. Unfortunately I don't think that can be changed on GKE (that's what we're using) so I'll have to upgrade my cluster to 1.23.

**Comment by alculquicondor** [comment]
The feature (the fix) is disabled in GKE in 1.23 because we found issues in the implementation. Stay tuned for the version where it will be enabled. 1.25 will for sure be fixed.

**Comment by alculquicondor** [comment]
>  In my case it looks like pods are getting deleted when the node gets removed (autoscaling).

Unfortunately, that case cannot be mitigated in 1.22-1.24 by changing terminated-pod-gc-threshold, because the Pod is removed immediately when the node is deleted.

**Comment by adriangb** [comment]
Makes sense.

> The feature (the fix) is disabled in GKE in 1.23 because we found issues in the implementation. Stay tuned for the 1.24 version where it will be enabled. 1.25 will for sure be fixed.

Is the "fix" the `JobTrackingWithFinalizers` feature? That is, do I need to do anything to get this working properly if I update to 1.24 or do I have to wait for 1.25?

**Comment by deenar** [comment]
This behavior of Kubernetes is down to the Kubernetes JobTracker relying on completed Pods to not be removed in order to track the Job completion status. See the [Kubernetes Enhancement Proposal (KEP-2307)](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/2307-job-tracking-without-lingering-pods) for more details.

What happens on the scale down of nodes, then completed pods are evicted, but the JobTracker relies on the existence of this Pod to derive the Job Completions status, viz Pods Statuses: 4 Running / 2 Succeeded / 0 Failed. After the pods are killed on a scale down, At the next JobSync the index completion get's reset back and all the Pods killed for the index are restarted.

Setting pod disruption budget or "cluster-autoscaler.kubernetes.io/safe-to-evict": "false" does not help as these pods are successful and marked ready for garbage collection. These are not used by the ClusterAutoScaler for auto-scaling decisions.

The fix is to turn on the feature gate [JobTrackingWithFinalizers](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#:%7E:text=Feature%20gates%20are%20a%20set,flag%20on%20each%20Kubernetes%20component) or upgrade to version 1.26 or later where this is the default. This enhancement removes the dependency on Pods being around for Job completion tracking. The Job controller creates Pods with a finalizer to prevent finished Pods from being removed by the garbage collector. The Job controller removes the finalizer from the finished Pods once it has accounted for them. It does this by keeping an additional state in the Job object. In subsequent Job syncs, the controller ignores finished Pods that don't have the finalizer

Follow the change tracking in this [issue](https://github.com/kubernetes/kubernetes/issues/108887)
